### DISCUSSION TITLE
Error Handling with ApplyT

### DISCUSSION THREAD

**Discussion by shreecodes** [pulumi/pulumi#9069] [Pulumi in Practice] [answered]
Hello 👋🏽 

We're using the Pulumi Golang sdk.

What's the conventional way to bubble up errors from within an `ApplyT`? I noticed that the function can return an error...but doing so does not seem to have any effect.

Are we supposed to call `panic(err)` within the interpolation function? And if so, is there a risk of that `panic` causing issues with the pulumi state or the infra state?

**Comment by jaxxstorm** [comment, marked as answer]
Hi there,

It depends really on how you're using the `ApplyT`

If you're assigning the result to a variable, the resulting variable will be a pulumi.Output. The `pulumi.Run` function returns an error, so you can do something like this:

```go
package main

import (
	"fmt"
	"strings"

	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {

		stringVar, err := random.NewRandomString(ctx, "string", &random.RandomStringArgs{
			Length: pulumi.Int(200),
		})
		if err != nil {
			return err
		}
		
		checkForLetter := stringVar.Result.ApplyT(func(value string) (string, error) {
			if strings.Contains(value, "a") {
				return value, nil
			} else {
				return "", fmt.Errorf("no letter a found in value")
			}
		})



		ctx.Export("letter", checkForLetter)

		return nil
	})
}
```

This accurately returns an error to the program, and there's no issue with affecting the state file here. You can't bubble up the resulting error to check for it, because inside an `ApplyT` is an eventual operation. 

If you want to exit a Pulumi program immediately, you can of course `panic` or return a `ctx.Log.Error`

**Reply by shreecodes** [reply]
Thanks for the clarification!

> This accurately returns an error to the program

Can you elaborate a little bit on this please? Does the returned output (`checkForLetter` in your example) get flagged as having an error internally?

I dug a bit into the sdk to try and find the answer myself. What does it mean for that state to be rejected? https://github.com/pulumi/pulumi/blob/master/sdk/go/pulumi/types.go#L472
